pm_tb_data/fetch/nrt.py

```diff
@@ -0,0 +1,186 @@
+        headers = {k.lower(): v for k, v in resp.headers.items()}
+        resp.close()
+
+        redirected = resp.status_code == 302
+        redirected_to_urs = (
+            redirected and "urs.earthdata.nasa.gov" in headers["location"]
+        )
+
+        if not (redirected_to_urs):
+            print(f"Host {host} did not redirect to URS -- continuing without auth.")
+            return s
+
+        auth_resp = s.get(
+            headers["location"],
+            # Don't download data!
+            stream=True,
+            auth=_get_earthdata_creds(),
+        )
+        resp.close()
+        if not (auth_resp.ok and s.cookies.get(_URS_COOKIE) == "yes"):
+            msg = f"Authentication with Earthdata Login failed with:\n{auth_resp.text}"
+            raise RuntimeError(msg)
+
+        print(f"Authenticated for {host} with Earthdata Login.")
+
+    return s
+
+
+FileType = Literal["R04", "P04"]
+
+
+class GranuleInfo(TypedDict):
+    file_type: FileType
+    filename: str
+    data_url: str
+
+
+GranuleInfoByDate = dict[dt.date, GranuleInfo]
+
+
+def _get_granule_info_by_date(*, data_granules: list[DataGranule]) -> GranuleInfoByDate:
+    fn_pattern = re.compile(
+        r"AMSR_U2_L3_SeaIce12km_(?P<file_type>P04|R04)_(?P<file_date>\d{8}).he5"
+    )
+    # TODO: better name/data structure.
+    granules_by_date: GranuleInfoByDate = {}
+    for granule in data_granules:
+        # The `native-id` of each granule is the filename. E.g.,
+        # `AMSR_U2_L3_SeaIce12km_R04_20230930.he5`.
+        filename = granule["meta"]["native-id"]
+        if not (match := fn_pattern.match(filename)):
+            # TODO: custom `pm_tb_data` error
+            raise FetchRemoteDataError(
+                "Found unexpected filename in CMR results (`native-id`)"
+                f": {filename}."
+            )
+        file_type = match.group("file_type")
+        file_type = cast(FileType, file_type)
+        file_date_str = match.group("file_date")
+        file_date = dt.datetime.strptime(file_date_str, "%Y%m%d").date()
+        # There are two links for each granule. one for lance.nsstc.nasa.gov and
+        # the other for lance.itsc.uah.edu. The first one is fine.
+        data_url = granule.data_links(access="external")[0]
+        granules_by_date[file_date] = {
+            "file_type": file_type,
+            "filename": filename,
+            "data_url": data_url,
+        }
+
+    return granules_by_date
+
+
+def _filter_out_last_day(*, granules_by_date: GranuleInfoByDate) -> GranuleInfoByDate:
+    """Remove the last day of data, unless it is an R04 file."""
+    filtered_granules_by_date = copy.deepcopy(granules_by_date)
+    dates = sorted(granules_by_date.keys())
+    # If the latest date is a partial file, discard it. We don't trust any data
+    # files earlier than the second-to-latest, unless the latest is an R04 file.
+    if granules_by_date[dates[-1]]["file_type"] == "P04":
+        del filtered_granules_by_date[dates[-1]]
```

Comment:
So...there is a slight bit of logic to discuss here.  If there is no file at all for a given day -- as happened on Sep 2, 2023 -- then on the following day, the list here will be insufficient.

The date for which we are able to compute data is "the date prior to the latest date in the list of file dates", not "the latest date in the list if it's an 'R' file, or the latest file date after the latest 'P' date has been removed."

Reply:
We discussed this:

We'll keep this logic as-is. From the `fetch` code's point of view, the only important thing is to make sure we don't data files that are subject to change to the local disk. We always avoid downloading the latest date of data _unless_ it's an `R` file.

In the `seaice_ecdr` code, we'll add logic to handle the case where there's missing data for a day of processing (e.g,. if it's Sept. 4  and no data for Sept. 3 are available, then the ECDR will temporally interpolate Sept. 3.)